### PR TITLE
Code coverage runtime

### PR DESCRIPTION
**Pull request by myrrc**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):

- Build/Testing/Packaging Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Сode coverage runtime for functional tests (replaces coverage CI check).
Uses clang SanitizerCoverage callbacks to store data on a per-test basis and dump them after testing.
Introduces critical edges coverage metric.

### CODE REVIEW

**Comment by myrrc** [comment]
As for now, building with -DENABLE_LIBRARIES=OFF produces such segdfault:

```
#0  0x0000000009a94fcd in dlerror () at ../base/harmful/harmful.c:199
#1  0x00000000099b28b4 in __ubsan::InitAsStandalone() ()
#2  0x0000000009997746 in __ubsan::PreInitAsStandalone() ()
#3  0x00007ffff7de39a6 in _dl_init (main_map=0x7ffff7ffe170, argc=1,
    argv=0x7fffffffe4f8, env=0x7fffffffe508) at dl-init.c:104
#4  0x00007ffff7dd40ca in _dl_start_user () from /lib64/ld-linux-x86-64.so.2
#5  0x0000000000000001 in ?? ()
#6  0x00007fffffffe72c in ?? ()
#7  0x0000000000000000 in ?? ()
```

Without this flag, 

```
#0  0x00007ffff74e7d34 in __GI___clock_gettime (clock_id=6, tp=0x7fffffffc760)
    at ../sysdeps/unix/clock_gettime.c:115
#1  0x00000000182cf9e5 in nstime_get (time=time@entry=0x7ffff6e06eb8)
    at ../contrib/jemalloc/src/nstime.c:119
#2  0x00000000182cf99d in nstime_update_impl (time=0x7ffff6e06eb8)
    at ../contrib/jemalloc/src/nstime.c:160
#3  0x0000000018296982 in arena_decay_reinit (decay=decay@entry=0x7ffff6e06df8,
    decay_ms=<optimized out>, decay_ms@entry=10000)
    at ../contrib/jemalloc/src/arena.c:672
#4  0x0000000018295f87 in arena_decay_init (decay=decay@entry=0x7ffff6e06df8,
    decay_ms=10000, stats=stats@entry=0x7ffff6e009f0)
    at ../contrib/jemalloc/src/arena.c:693
#5  0x0000000018295cc1 in arena_new (tsdn=tsdn@entry=0x0, ind=0, ind@entry=144017864,
    extent_hooks=extent_hooks@entry=0x7fffffffc860)
    at ../contrib/jemalloc/src/arena.c:2043
#6  0x000000001827fbce in arena_init_locked (tsdn=<optimized out>, tsdn@entry=0x0,
    ind=<optimized out>, ind@entry=0, extent_hooks=<optimized out>,
    extent_hooks@entry=0x89589c8 <extent_hooks_default>)
    at ../contrib/jemalloc/src/jemalloc.c:340
#7  0x000000001827f888 in arena_init (tsdn=tsdn@entry=0x0, ind=ind@entry=0,
    extent_hooks=0x0) at ../contrib/jemalloc/src/jemalloc.c:368
#8  0x0000000018286a24 in malloc_init_hard_a0_locked ()
    at ../contrib/jemalloc/src/jemalloc.c:1539
#9  0x000000001828c2c7 in malloc_init_hard () at ../contrib/jemalloc/src/jemalloc.c:1736
#10 0x000000001828a2ca in malloc_init () at ../contrib/jemalloc/src/jemalloc.c:223
#11 imalloc_init_check (sopts=<optimized out>, dopts=dopts@entry=0x7fffffffe370)
    at ../contrib/jemalloc/src/jemalloc.c:2215
#12 0x0000000018280c52 in imalloc (sopts=sopts@entry=0x7fffffffe328,
    dopts=dopts@entry=0x7fffffffe370) at ../contrib/jemalloc/src/jemalloc.c:2246
#13 0x0000000018281295 in calloc (num=num@entry=1, size=size@entry=32)
#14 0x00007ffff79c87f5 in _dlerror_run (
    operate=operate@entry=0x7ffff79c80d0 <dlsym_doit>, args=args@entry=0x7fffffffe3f0)
    at dlerror.c:140
#15 0x00007ffff79c8166 in __dlsym (handle=<optimized out>,
    name=0x7ff1c3f "swift_demangle") at dlsym.c:70
#16 0x000000000e67018c in __sanitizer::Symbolizer::LateInitialize() ()
#17 0x000000000e672c64 in __ubsan::InitAsStandalone() ()
#18 0x000000000e657af6 in __ubsan::PreInitAsStandalone() ()
#19 0x00007ffff7de39a6 in _dl_init (main_map=0x7ffff7ffe170, argc=2,
    argv=0x7fffffffe4c8, env=0x7fffffffe4e0) at dl-init.c:104
#20 0x00007ffff7dd40ca in _dl_start_user () from /lib64/ld-linux-x86-64.so.2
#21 0x0000000000000002 in ?? ()
#22 0x00007fffffffe709 in ?? ()
#23 0x00007fffffffe743 in ?? ()
#24 0x0000000000000000 in ?? ()

```

**Comment by myrrc** [comment]
Looks like it's the `harmful` contrib that was causing the bugs, removed it from the coverage builds as we won't need it anyway.

**Comment by myrrc** [comment]
Looks like clang's whitelist/blacklist for SanitizerCoverage are completely broken -- I couldn't make them work in any way:
1) Providing a relative path for black/white lists.
2) Providing an absolute path (also, the hardcoded version).
3) Providing a mangled/demangled name.

The docs stated one could use a shell-like asterisk (*) expansion, but this didn't succeed either, so I decided to switch to explicit flags (-fsanitize-coverage and -fno-sanitize-coverage) on a per-cmake-target basis.

**Comment by myrrc** [comment]
A word about the IPC between the tester script and the server -- currently we have 3 main options:
1) Utilizing the tester's --testname option which executes a SELECT testname query before running the test.
The issue here is that we need to interfere with the query process pipeline -- either write a hook or handle the special coverage case with a macro.

2) Setting some option like SET coverage_test_name = name; Same issue with the pipeline.

3) Sending a signal from the tester to the server. This option is "cheaper" in a way, the only issue here is how to attach the test_id to the signal. The server uses the POSIX signal handler with the SA_SIGINFO. The tester preloads libc and executes the sigqueue function.

**Comment by myrrc** [comment]
First measurements.

About 560k basic block critical edges.

Build (minimal config):
real    `11m59.476s` (clean) vs `12m27.273s`
user   ` 180m58.127s` vs  `186m45.822s`
sys     `7m22.119s` vs `7m38.237s`.

Binary size:
`1659MB` (clean) vs `1699MB`

**Comment by myrrc** [comment]
The first more-or-less working version is here. The slowdown for test processing (the data collection step only) is about 10x (0.2s -> 2s) or so. The test conversion takes about 10m for the test (but I guess the global symbols cache will speedup is, so it's not 10m for every test, but only for the first one).


[test_out.info.txt -- lCOV report, used for HTML generation](https://github.com/ClickHouse/ClickHouse/files/6405829/test_out.info.txt).

[time_profile.lz.gz (warning: GitHub does not accept .lz archives, so it's uploaded as .gz, but the real extension is .lz)](https://github.com/ClickHouse/ClickHouse/files/6405979/time_profile.lz.gz).


Generated report, v0, top view
Generated report, v0, file view

**Comment by myrrc** [comment]
Measures with full build:

- Testing takes about 4.88h (328 errors! 2358 tests passed. 300 tests skipped. 17555.63 s elapsed), but 
    - CI has more powerful hardware (so we could increase the thread pool size from 16 (current) to 24 for initial symbolization and from 10 to 15 for parallel test dumping).
    - Some tests failed with a maximum timeout (180s) as they maintain invariants i.e. "clickhouse's data is located in /var/lib/clickhouse" which were invalid for my local testing.
- gzipped data takes 2.8gb on the drive (~1.1mb per test if storing actual .info data) or about 300mb (~92kb per test if storing gzipped diffs with the first test).
- `genhtml` is certainly the bottleneck here, and sadly we can't build per-test incremental code coverage as it's too slow.
[genhtml_log.log](https://github.com/ClickHouse/ClickHouse/files/6428536/genhtml_log.log)

Current proposed solution is to accumulate global coverage information and dump it (as a global report) after all tests end.
Then we can build two reports with genhtml -- absolute accumulated coverage for commit and incremental coverage (would use previous report to highlight changes in code coverage).

**Comment by myrrc** [comment]
After introducing own memory pool I managed to lower execution time to 14642.36 s ~ 40.2h

**Comment by myrrc** [comment]
Last commit https://github.com/ClickHouse/ClickHouse/pull/20539/commits/2a08dd3f671b154268a892045a290e426baeaddc introduced a working version with dumping genhtml reports, but this format still isn't appropriate for accumulating large coverage reports while preserving per-test data, so I'll create my own format and write a converter to genhtml.

**Comment by myrrc** [comment]
Own format reduced testing time to about 3h 23 minutes.

**Comment by myrrc** [comment]
Runtime works locally but does not work in container.
Currently, there is an eternal lock in Coverage.cpp:74 while waiting to join the first thread.
Hovewer, at this time point the shutdown variable is set to true, and all threads' condition varables were notified of it, so they should at least start their jobs.

lldb also showed that at given time point the jobs queue consisted of 16 threads (hardware concurrency) which means no thread started its job (but each time a job is scheduled the schedule function notifies a condition variable so that's also weird).

**Comment by myrrc** [comment]
Added custom front:
image
image
image

File: base/ext/push_back.h [hunk 1]
```diff
@@ -4,22 +4,17 @@
...
 {
-
 /// Moves all arguments starting from the second to the end of the vector.
 /// For example, `push_back(vec, a1, a2, a3)` is a more compact way to write
 /// `vec.push_back(a1); vec.push_back(a2); vec.push_back(a3);`
 /// This function is like boost::range::push_back() but works for noncopyable types too.
-template <typename T>
-void push_back(std::vector<T> &)
+template <class T, class ...Args>
+void push_back(std::vector<T> & vec, Args&& ...args)
```
Comment (kitaisreal) [inline base/ext/push_back.h:12]: This method is no longer be presented in master because we will remove ext, consider to replace to some local function.

File: base/harmful/CMakeLists.txt [hunk 2]
```diff
@@ -1,2 +1,5 @@
-add_library(harmful harmful.c)
-install(TARGETS harmful EXPORT global ARCHIVE DESTINATION lib)
+# Some strage segfaults if linked
+if(NOT WITH_COVERAGE)
+    add_library(harmful harmful.c)
```
Comment (kitaisreal) [inline base/harmful/CMakeLists.txt:3]: If there are segfaults because of harmful library, that means that some code is using harmful function consider to fix it instead of disabling.

File: blocklist.txt [hunk 3]
```diff
@@ -0,0 +1,47 @@
+# There's no way to portably mangle the type name, so it's just for x86_64
+# Place here files that are instrumented even when providing noinstrument flags
+#
+# Global variables init for iostream
```
Comment (kitaisreal) [inline blocklist.txt:4]: This should be fixed before merge, with some better workaround.

File: cmake/linux/default_libs.cmake [hunk 4]
```diff
@@ -39,7 +39,10 @@ find_package(Threads REQUIRED)
 if (NOT OS_ANDROID)
     # Our compatibility layer doesn't build under Android, many errors in musl.
     add_subdirectory(base/glibc-compatibility)
-    add_subdirectory(base/harmful)
+
+    if (NOT WITH_COVERAGE)
+        add_subdirectory(base/harmful)
```
Comment (kitaisreal) [inline cmake/linux/default_libs.cmake:44]: Try to fix harmful usage with coverage.

File: contrib/jemalloc-cmake/CMakeLists.txt [hunk 5]
```diff
@@ -11,6 +11,10 @@ else ()
     option (ENABLE_JEMALLOC "Enable jemalloc allocator" ${ENABLE_LIBRARIES})
 endif ()
 
+if (WITH_COVERAGE)
+    set (ENABLE_JEMALLOC ON)
```
Comment (kitaisreal) [inline contrib/jemalloc-cmake/CMakeLists.txt:15]: Why for coverage we enable jemalloc ?

File: docker/test/coverage/ccr_converter.py [hunk 6]
```diff
@@ -0,0 +1,517 @@
...
+
+    read_tests(report_file, source_files_map)
+
+    tests_names = list(map(lambda x: x[:-1], report_file.readlines())) # remove \n
+
+    assert len(tests_names) == len(tests), f"{len(tests_names)} != {len(tests)}"
+
+    print(f"Read the report, {len(tests)} tests, {len(files)} source files")
+
+def main():
```
Comment (kitaisreal) [inline docker/test/coverage/ccr_converter.py:488]: Add comments about this file.

File: programs/CMakeLists.txt [hunk 7]
```diff
@@ -322,11 +322,12 @@ else ()
     add_executable (clickhouse main.cpp)
 
     # A library that prevent usage of several functions from libc.
-    if (ARCH_AMD64 AND OS_LINUX AND NOT OS_ANDROID)
+    if (ARCH_AMD64 AND OS_LINUX AND NOT OS_ANDROID AND NOT WITH_COVERAGE)
         set (HARMFUL_LIB harmful)
```
Comment (kitaisreal) [inline programs/CMakeLists.txt:326]: Try to fix harmful lib usage with coverage.

File: programs/client/CMakeLists.txt [hunk 8]
```diff
@@ -5,6 +5,15 @@ set (CLICKHOUSE_CLIENT_SOURCES
     Suggest.cpp
 )
 
+if (WITH_COVERAGE)
+    set_source_files_properties(
+        Client.cpp
+        ConnectionParameters.cpp
+        QueryFuzzer.cpp
+        Suggest.cpp
+        PROPERTIES COMPILE_FLAGS "${WITHOUT_COVERAGE}")
```
Comment (kitaisreal) [inline programs/client/CMakeLists.txt:14]: Add comment why this is compiled without coverage.

File: programs/client/Client.cpp [hunk 9]
```diff
@@ -2694,6 +2698,10 @@ class Client : public Poco::Util::Application
 
 int mainEntryClickHouseClient(int argc, char ** argv)
 {
+#if WITH_COVERAGE
+    ::detail::Writer::instance().onClientInitialized();
```
Comment (kitaisreal) [inline programs/client/Client.cpp:2702]: Remove Writer from detail namespace, detail namespaces should not be used by clients. Consider just Writer::instance().

File: src/Common/Coverage.cpp [hunk 10]
```diff
@@ -0,0 +1,582 @@
...
+      dwarf()
+#else
+      symbol_index(getInstanceAndInitGlobalCounters()),
+      dwarf(symbol_index->getSelf()->elf)
+#endif
+    {}
+
+Writer& Writer::instance()
+{
+    /// Meyer's singleton.
```
Comment (kitaisreal) [inline src/Common/Coverage.cpp:106]: Comments like this are not needed.

File: src/Common/ProfileEvents.cpp [hunk 11]
```diff
@@ -258,6 +258,7 @@ constexpr Event END = __COUNTER__;
 
 /// Global variable, initialized by zeros.
 Counter global_counters_array[END] {};
+
```
Comment (kitaisreal) [inline src/Common/ProfileEvents.cpp:261]: Why there are unrelated changes ? Do we have unnecessary spaces here ?

File: src/Common/Coverage.h [hunk 12]
```diff
@@ -0,0 +1,272 @@
...
+    {
+        FILE * handle {nullptr};
+    public:
+        inline void set(const std::string& pathname, const char * mode) { handle = fopen(pathname.data(), mode); }
+        inline FILE * file() { return handle; }
+        inline void close() { fclose(handle); handle = nullptr; }
+        inline void write(const fmt::memory_buffer& mb) { fwrite(mb.data(), sizeof(char), mb.size(), handle); }
+    };
+
+    FileWrapper report_file;
```
Comment (kitaisreal) [inline src/Common/Coverage.h:241]: Why we use custom FileWrapper implementation, cant we use WriteBufferFromFile ?

File: src/Common/Coverage.h [hunk 13]
```diff
@@ -0,0 +1,272 @@
...
+ * - Does not use metrics.
+ * - Some operations are not thread-safe.
+ *
+ * Own implementation needed as Writer does not use most FreeThreadPool features (and we can save ~20 minutes by
+ * using this class instead of the former).
+ *
+ * Each test duration is longer than coverage test processing pipeline (converting and dumping to disk), so no
+ * more than 1 thread is needed.
+ */
+class TaskQueue
```
Comment (kitaisreal) [inline src/Common/Coverage.h:91]: How and why this is faster ? Why cant we create our thread pool with one thread ?

File: src/Common/Coverage.h (hunk 12 above)
Comment (myrrc) [reply src/Common/Coverage.h:241]: No, otherwise we would need to exclude this file from instrumentation.

File: contrib/jemalloc-cmake/CMakeLists.txt (hunk 5 above)
Comment (myrrc) [reply contrib/jemalloc-cmake/CMakeLists.txt:15]: Faster test runs for contribs.

File: programs/client/Client.cpp (hunk 9 above)
Comment (myrrc) [reply programs/client/Client.cpp:2702]: This code is not intended to be used by clients (only by coverage runtime developer), so internal namespace is OK as it emphasises runtime's task.

File: src/Common/CoverageCallbacks.h [hunk 14]
```diff
@@ -0,0 +1,24 @@
+#pragma once
+#include <numeric>
+#include "Coverage.h"
+
+//NOLINTNEXTLINE(bugprone-reserved-identifier, readability-non-const-parameter)
+extern "C" void __sanitizer_cov_trace_pc_guard_init(uint32_t *start, uint32_t *stop)
+{
+    if (start == stop || *start) return;
+    std::iota(start, stop, 1);
```
Comment (kitaisreal) [inline src/Common/CoverageCallbacks.h:9]: Why we fill it with 1 ?

File: src/Common/Dwarf.h [hunk 15]
```diff
@@ -127,15 +127,10 @@ class Dwarf final
...
-        // Don't resolve location info.
-        DISABLED,
-        // Perform CU lookup using .debug_aranges (might be incomplete).
-        FAST,
-        // Scan all CU in .debug_info (slow!) on .debug_aranges lookup failure.
-        FULL,
-        // Scan .debug_info (super slower, use with caution) for inline functions in
-        // addition to FULL.
-        FULL_WITH_INLINE,
+        DISABLED, /// Don't resolve location info.
```
Comment (kitaisreal) [inline src/Common/Dwarf.h:130]: It seems in our places we use comments on top, why here we need to place them right ?

File: src/Common/ProfileEvents.h [hunk 16]
```diff
@@ -20,6 +20,7 @@ namespace ProfileEvents
 
     /// Counters - how many times each event happened
     extern Counters global_counters;
+    extern Counter global_counters_array[]; //needed for Coverage.cpp
```
Comment (kitaisreal) [inline src/Common/ProfileEvents.h:23]: Why it is needed for coverage ?

File: src/Common/SymbolIndex.h [hunk 17]
```diff
@@ -43,6 +45,9 @@ class SymbolIndex : private boost::noncopyable
     const Symbol * findSymbol(const void * address) const;
     const Object * findObject(const void * address) const;
 
+    /// Retrieves the clickhouse binary object.
+    inline const Object * getSelf() const { return findObject(&dummy_var); }
```
Comment (kitaisreal) [inline src/Common/SymbolIndex.h:49]: Add comment why this works.

File: src/Common/CoverageCallbacks.h (hunk 14 above)
Comment (myrrc) [reply src/Common/CoverageCallbacks.h:9]: See sanitizer documentation https://clang.llvm.org/docs/SanitizerCoverage.html#id2

File: src/Interpreters/Context.cpp [hunk 18]
```diff
@@ -1111,12 +1115,16 @@ void Context::setSettings(const Settings & settings_)
...
+    const std::string_view name_view {name};
+
     if (name == "profile")
     {
         setProfile(value);
         return;
     }
-    settings.set(std::string_view{name}, value);
+
+    settings.set(name_view, value);
```
Comment (kitaisreal) [inline src/Interpreters/Context.cpp:1127]: Why this is changed ?

**Review by kitaisreal — CHANGES_REQUESTED** [review]
Need to address questions.

File: src/Common/Coverage.h (hunk 13 above)
Comment (myrrc) [reply src/Common/Coverage.h:91]: We can't use internal CH pools as they would need to be excluded from instrumentation (due to runtime implementation).
TaskQueue _is_ a thread pool with one thread, the only difference is waiting on condition variable.
Motivation for that is clear in the comment:
>  Each test duration is longer than coverage test processing pipeline (converting and dumping to disk)

File: src/Common/Dwarf.h (hunk 15 above)
Comment (myrrc) [reply src/Common/Dwarf.h:130]: Multiline right-padded comments are more ugly IMO.

File: src/Common/SymbolIndex.h (hunk 17 above)
Comment (myrrc) [reply src/Common/SymbolIndex.h:49]: It's quite obvious that static variables are located inside binary/DSO that contains translation unit where such variable is declared.

**Comment by myrrc** [comment]
Depends on  https://github.com/ClickHouse/ClickHouse/pull/27228

**Comment by myrrc** [comment]
Depends on https://github.com/ClickHouse/ClickHouse/pull/27361

**Comment by myrrc** [comment]
I think this PR won't be done in near future, so I'm closing it.

If you operate entirely on the information provided by the binary itself (with or without PC-table), all you can get is addresses that were hit (or some basic blocks indices that were hit). You can symbolize this data and get the source file and line, but the problem is that, you can't really do anything with this information.

Multiple basic blocks can correspond to some lines (e.g. function template instantiations). Basic blocks can correspond to lines not belonging to functions (e.g. macro expansion like settings traits implementation).
If you instrument the code at critical edges level, you lose the access to basic blocks structure, these addresses (of edges) could as well be generated randomly as there's no profit from using them.
You won't be able to get lines ranges that were hit as C++ is hard to parse and you need a compiler.

If you involve the compiler (.gcno), things don't get better. clang does not enumerate its basic blocks, so there's no way to match the basic block that was hit, for example, in the boolean counters array (the PC-table) with a basic block parsed from a gcno file. You get stuck at the function level. Sorting-transforming basic blocks line ranges still does not solve the problem (moreover, .gcno produces strange results like a line belonging to a basic block that's empty). gcno parsers like lcov use perl scripts containing thousands of lines to turn all this information into a human-readable format.

And if one day you want to try the source-based code coverage, you'll also fail. There's no way to tweak what data are stored and tracked (unless you patch the compiler which obviously is not a good idea here), so you just collect everything (and it takes >= 9 hours currently for a general run without per-test data). All you can do is use builtins provided to write the report to some other location, but it won't make everything substantially faster.

You may wonder: why can't you simply collect addresses that were hit and display them? Well, for CH purposes it's a) useless, as one needs to get the coverage percentage, and b) useless, as you reinvent the sancov wheel.

I doubt there's a good way to resolve this PR.

**Comment by alexey-milovidov** [comment]
> You may wonder: why can't you simply collect addresses that were hit and display them?

I'm going to do exactly this in https://github.com/ClickHouse/ClickHouse/pull/56102

> useless, as one needs to get the coverage percentage

The percentage can be calculated as the percentage of all instrumented edges, the percentage of instrumented symbols in the binary, functions, or source files. The percentage of lines can be calculated roughly if you take the assumption that every basic block spans from the line corresponding to its address to the next instrumented line.
